numpy/_core/src/multiarray/stringdtype/static_string.c

```diff
@@ -0,0 +1,745 @@
+        f(allocator->arena.buffer);
+    }
+    if (allocator->allocator_lock != NULL) {
+        PyThread_free_lock(allocator->allocator_lock);
+    }
+
+    f(allocator);
+}
+
+static int
+is_short_string(const npy_packed_static_string *s)
+{
+    unsigned char high_byte =
+            ((_npy_static_string_u *)s)->direct_buffer.size_and_flags;
+    int has_short_flag = (high_byte & NPY_STRING_SHORT);
+    int has_on_heap_flag = (high_byte & NPY_STRING_ON_HEAP);
+    return has_short_flag && !has_on_heap_flag;
+}
+
+static int
+is_medium_string(const _npy_static_string_u *s)
+{
+    unsigned char high_byte = s->direct_buffer.size_and_flags;
+    int has_short_flag = (high_byte & NPY_STRING_SHORT);
+    int has_medium_flag = (high_byte & NPY_STRING_MEDIUM);
+    return (!has_short_flag && has_medium_flag);
+}
+
+NPY_NO_EXPORT int
+NpyString_isnull(const npy_packed_static_string *s)
+{
+    unsigned char high_byte =
+            ((_npy_static_string_u *)s)->direct_buffer.size_and_flags;
+    return (high_byte & NPY_STRING_MISSING) == NPY_STRING_MISSING;
+}
+
+static int
+is_not_a_vstring(const npy_packed_static_string *s)
+{
+    return is_short_string(s) || NpyString_isnull(s);
+}
+
+static int
+is_a_vstring(const npy_packed_static_string *s)
+{
+    return !is_not_a_vstring(s);
+}
+
+/*NUMPY_API
+ * Acquire the mutex locking the allocator attached to *descr*.
+ *
+ * NpyString_release_allocator must be called on the allocator returned
+ * by this function exactly once.
+ */
+NPY_NO_EXPORT npy_string_allocator *
+NpyString_acquire_allocator(PyArray_StringDTypeObject *descr)
+{
+    if (!PyThread_acquire_lock(descr->allocator->allocator_lock, NOWAIT_LOCK)) {
+        PyThread_acquire_lock(descr->allocator->allocator_lock, WAIT_LOCK);
+    }
+    return descr->allocator;
+}
+
+/*NUMPY_API
+ * Simultaneously acquire the mutexes locking the allocators attached to
+ * multiply descriptors.
+ *
+ * Note that the descrs array can have more elements than n_allocators, and
+ * any non-stringdtype descriptors are skipped, allowing passing the
+ * `context->descriptors` array inside a ufunc or cast loop.
+ *
+ * There is no error checking for setting n_allocators too large and buffer
+ * overflows might happen if there are not n_allocators StringDType descriptors
+ * in the descrs array.
+ *
+ * If pointers to the same descriptor are passed multiple times, only acquires
+ * the allocator mutex once but sets identical allocator pointers appropriately.
+ *
+ * The allocator mutexes must be released after this function returns, see
+ * NpyString_release_allocators.
+ */
+NPY_NO_EXPORT void
+NpyString_acquire_allocators(size_t n_allocators,
+                             PyArray_Descr *descrs[],
+                             npy_string_allocator *allocators[])
+{
+    size_t allocator_count = 0;
+    for (size_t i=0; allocator_count<n_allocators; i++) {
+        if (NPY_DTYPE(descrs[i]) != &PyArray_StringDType) {
+            continue;
```

Comment:
No, this function returns an array that is n_allocators long unless it can't find n_allocators descriptors to fill the array with. The `continue` and not setting NULL is on purpose. Also note how I'm setting `allocators[allocator_count]`, not `allocators[i]` below. I don't want to have any NULL entires in the `allocators` array gets filled out so the callers don't have to keep track of that.

Reply:
There's only one spot where I pass in an extra descriptor, in replace: https://github.com/ngoldbaum/numpy/blob/6263ea039665b17c9d6fdb4f33668b9c3647536d/numpy/_core/src/umath/stringdtype_ufuncs.cpp#L2472-L2477

This API is the nicest possibility for the caller IMO because they can just pass `context->descriptors` without processing it. They do need to know the number of input stringdtype descriptors, but you know that from the definition of the ufunc. After the ufunc loop, they just need to pass the `allocators` array to `NpyString_release_allocator`, again without any processing and with a statically knowable length: https://github.com/ngoldbaum/numpy/blob/6263ea039665b17c9d6fdb4f33668b9c3647536d/numpy/_core/src/umath/stringdtype_ufuncs.cpp#L2552-L2557